crypto/rand/rand_unix.c

```diff
@@ -231,23 +229,34 @@ static size_t sysctl_random(char *buf, size_t buflen)
  */
 int syscall_random(void *buf, size_t buflen)
 {
-#  if defined(OPENSSL_HAVE_GETRANDOM)
-    return (int)getrandom(buf, buflen, 0);
-#  endif
+    static union {
+        void *p;
+        int (*f)(void *buffer, size_t length);
+    } p_getentropy = { NULL };
+
+    /*
+     * Do runtime detection to find getentropy().
+     *
+     * Known OSs that should support this:
+     * - Darwin since 16 (OSX 10.12, IOS 10.0).
+     * - Solaris since 11.3
+     * - OpenBSD since 5.6
+     * - Linux since 3.17 with glibc 2.25
+     * - FreeBSD since 12.0 (1200061)
+     */
+    p_getentropy.p = DSO_global_lookup("getentropy");
```

Comment:
Do you actually mean to invoke DSO_global_lookup for each invocation of syscall_random? It can be quite expensive. It's more appropriate to initialize it only once, but then value has to be "tri-state": uninitialized, available, and not available. Minor challenge is to make it MT-safe. One doesn't need lock, because DSO_global_lookup will return same value, but one has to ensure that there is only one store.

Reply:
Then it would be appropriate to remove `static` p_getentropy declaration and drop a comment. No?

Reply:
Oh, I didn't know I copied that static, let me fix that.